### PR TITLE
Expose source_model parameter for vector-enabled collections

### PR DESCRIPTION
**Pull request by Hazel-Datastax**
**What this PR does**:
Expose source_model parameter for vector-enabled collections

**Which issue(s) this PR fixes**:
Fixes [JiraC2-3495](https://datastax.jira.com/browse/C2-3495)

**Checklist**
- [x] Changes manually tested
- [x] Automated Tests added/updated
- [ ] Documentation added/updated
- [x] CLA Signed: [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

File: src/main/java/io/stargate/sgv2/jsonapi/service/schema/collections/CollectionSchemaObject.java [hunk 1]
```diff
@@ -274,27 +294,12 @@ private static CollectionSchemaObject createCollectionSettings(
...
-        SimilarityFunction.fromString(jsonNode.get("metric").asText());
-
-    return VectorConfig.ColumnVectorDefinition.fromJson(
-        DocumentConstants.Fields.VECTOR_EMBEDDING_TEXT_FIELD,
-        dimension,
-        similarityFunction,
-        jsonNode,
-        objectMapper);
-  }
-
```
Comment (Hazel-Datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/schema/collections/CollectionSchemaObject.java:297]: Remove the unused method. It's been removed to `VectorConfig` class

File: src/main/java/io/stargate/sgv2/jsonapi/config/constants/VectorConstant.java [hunk 2]
```diff
@@ -13,17 +13,17 @@ public interface VectorConstant {
...
           "gecko",
           SimilarityFunction.DOT_PRODUCT,
-          "nv_qa_4",
+          "nv-qa-4",
           SimilarityFunction.DOT_PRODUCT,
-          "cohere_v3",
+          "cohere-v3",
           SimilarityFunction.DOT_PRODUCT,
           "other",
           SimilarityFunction.COSINE);
```
Comment (Hazel-Datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/config/constants/VectorConstant.java:29]: According to the [document](https://docs.datastax.com/en/astra-db-serverless/cql/query-vector-data-with-cql.html#create-the-vector-schema), here should be hyphens not underscores. Since changing the value in here will affect the code in tables. I will create another PR to replace those hardcoded values to Emun.

File: src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CreateCollectionCommandResolver.java [hunk 3]
```diff
@@ -192,13 +197,35 @@ private CreateCollectionCommand.Options.VectorSearchConfig validateVectorOptions
...
+    }
 
     if (service != null) {
       // Validate service configuration and auto populate vector dimension.
       vectorDimension = validateVectorize.validateService(service, vectorDimension);
       vector =
           new CreateCollectionCommand.Options.VectorSearchConfig(
-              vectorDimension, vector.metric(), vector.vectorizeConfig());
+              vectorDimension, metric, sourceModel, vector.vectorizeConfig());
     } else {
```
Comment (Hazel-Datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/resolver/CreateCollectionCommandResolver.java:229]: Do we want to optimize the `sourceModel` for service? For example, if the user only specifies `openai` and `text-embedding-3-small` in the `service`, do we want to optimize the `sourceModel` to `openai-v3-small`. Currently, we will use `other` as the default value.

**Review by vkarpov15 — COMMENTED** [review]
LGTM

File: src/main/java/io/stargate/sgv2/jsonapi/api/configuration/CommandObjectMapperHandler.java [hunk 4]
```diff
@@ -38,7 +38,7 @@ public boolean handleUnknownProperty(
     }
     if (typeStr.endsWith("CreateCollectionCommand$Options$VectorSearchConfig")) {
       throw ErrorCodeV1.INVALID_CREATE_COLLECTION_OPTIONS.toApiException(
-          "Unrecognized field \"%s\" for `createCollection.options.vector` (known fields: \"dimension\", \"metric\", \"service\")",
+          "Unrecognized field \"%s\" for `createCollection.options.vector` (known fields: \"dimension\", \"metric\", \"sourceModel\", \"service\")",
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/configuration/CommandObjectMapperHandler.java:41]: nit: keep property names in alphabetic order, so `... \"service\", \"sourceModel\")"`

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/impl/CreateCollectionCommand.java [hunk 5]
```diff
@@ -95,6 +95,20 @@ public record VectorSearchConfig(
             @JsonProperty("metric")
             @JsonAlias("function") // old name
             String metric,
+        @Nullable
+            @Pattern(
+                regexp =
+                    "(openai-v3-large|openai-v3-small|ada002|gecko|nv-qa-4|cohere-v3|bert|other)",
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/impl/CreateCollectionCommand.java:101]: Should these also be in alphabetic order (except for "other" that should be last I think)?

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/impl/CreateCollectionCommand.java [hunk 6]
```diff
@@ -105,9 +119,11 @@ public record VectorSearchConfig(
             @JsonProperty("service")
             VectorizeConfig vectorizeConfig) {
 
-      public VectorSearchConfig(Integer dimension, String metric, VectorizeConfig vectorizeConfig) {
+      public VectorSearchConfig(
+          Integer dimension, String metric, String sourceModel, VectorizeConfig vectorizeConfig) {
         this.dimension = dimension;
-        this.metric = metric == null ? "cosine" : metric;
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/api/model/command/impl/CreateCollectionCommand.java:110]: Is this change intentional wrt metric? (I assume so but since there is no comment thought I'd ask)

File: src/main/java/io/stargate/sgv2/jsonapi/config/constants/VectorConstant.java [hunk 7]
```diff
@@ -13,17 +13,17 @@ public interface VectorConstant {
       Map.of(
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/config/constants/VectorConstant.java:13]: nit: should probably be `SUPPORTED_SOURCE_MODELS` (instead of `SUPPORTED_SOURCE_MODELS`).

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java [hunk 8]
```diff
@@ -141,6 +141,8 @@ public enum ErrorCodeV1 {
 
   VECTOR_SEARCH_INVALID_FUNCTION_NAME("Invalid vector search function name"),
 
+  VECTOR_SEARCH_INVALID_SOURCE_MODEL_NAME("Invalid vector search source model name"),
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java:144]: Not sure but should we use "Unrecognized" instead of "Invalid" for this (I know we use "invalid" above so maybe it's more consistent)?

File: src/main/java/io/stargate/sgv2/jsonapi/service/schema/SourceModel.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  COHERE_V3("cohere-v3"),
+  GECKO("gecko"),
+  NV_QA_4("nv-qa-4"),
+  OPENAI_V3_LARGE("openai-v3-large"),
+  OPENAI_V3_SMALL("openai-v3-small"),
+  OTHER("other"),
+  UNDEFINED("undefined");
+
+  private final String sourceModel;
+  private static final java.util.Map<String, SourceModel> SOURCE_MODELS_MAP =
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/schema/SourceModel.java:23]: nit: usually we use `import` instead of fully-qualified name.

File: src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java (hunk 8 above)
Comment (Hazel-Datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/exception/ErrorCodeV1.java:144]: Yes...I used "invalid" because the above used it. Changed to "Unrecognized", should we change the above to "Unrecognized" as well?

File: src/main/java/io/stargate/sgv2/jsonapi/service/schema/SourceModel.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    return sourceModel;
+  }
+
+  public static SourceModel fromString(String sourceModel) {
+    if (sourceModel == null) return UNDEFINED;
+    // The string may be empty if the collection was created before supporting source models
+    if (sourceModel.isEmpty()) return OTHER;
+    SourceModel model = SOURCE_MODELS_MAP.get(sourceModel);
+    if (model == null) {
+      throw ErrorCodeV1.VECTOR_SEARCH_INVALID_SOURCE_MODEL_NAME.toApiException("'%s'", sourceModel);
```
[the remark this answers is not in the document]
Comment (Hazel-Datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/service/schema/SourceModel.java:41]: Agree. I followed the pattern in [SimilarityFunction](https://github.com/stargate/data-api/blob/2ebf2696cdefb83bc3a6916acbf84b2d9ffb0f79/src/main/java/io/stargate/sgv2/jsonapi/service/schema/SimilarityFunction.java#L39). Do we want to change it as well?

File: src/test/java/io/stargate/sgv2/jsonapi/api/v1/CreateCollectionIntegrationTest.java [hunk 11]
```diff
@@ -1751,6 +1752,275 @@ public void failWithWrongModelParameterType() {
...
+          .statusCode(200)
+          .body("$", responseIsError())
+          .body("errors[0].exceptionClass", is("JsonApiException"))
+          .body("errors[0].errorCode", is("COMMAND_FIELD_INVALID"))
+          .body(
+              "errors[0].message",
+              startsWith(
+                  "Request invalid: field 'command.options.vector.sourceModel' value \"invalidName\" not valid. Problem: sourceModel options are 'openai-v3-large', 'openai-v3-small', 'ada002', 'gecko', 'nv-qa-4', 'cohere-v3', 'bert', and 'other'."));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Hazel-Datastax) [reply src/test/java/io/stargate/sgv2/jsonapi/api/v1/CreateCollectionIntegrationTest.java:2022]: If passing JSON Object, there will be a deserialization error from `CreateCollectionCommand`. Do we want to use this error?
```
{
    "errors": [
        {
            "message": "Request invalid, cannot parse as JSON: Cannot deserialize value of type `java.lang.String` from Object value (token `JsonToken.START_OBJECT`)\n at [Source: (ByteArrayInputStream); line: 6, column: 32] (through reference chain: io.stargate.sgv2.jsonapi.api.model.command.impl.CreateCollectionCommand[\"options\"]->io.stargate.sgv2.jsonapi.api.model.command.impl.CreateCollectionCommand$Options[\"vector\"]->io.stargate.sgv2.jsonapi.api.model.command.impl.CreateCollectionCommand$Options$VectorSearchConfig[\"sourceModel\"])",
            "errorCode": "INVALID_REQUEST_NOT_JSON"
        }
    ]
}
```

File: src/main/java/io/stargate/sgv2/jsonapi/api/model/command/impl/CreateCollectionCommand.java (hunk 6 above)
Comment (Hazel-Datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/api/model/command/impl/CreateCollectionCommand.java:110]: Yes. The `metric` somehow depends on the `sourceModel` now. If the `sourceModel` is not `other` and the user doesn't give the value to `metric`. We need to assign `dot_product` to `metric` (I handle it in `CreateCollectionCommandResolver`). If we assign `cosine` here, we can't know whether the `cosine` is given by the user or by our default value.

File: src/main/java/io/stargate/sgv2/jsonapi/config/constants/VectorConstant.java (hunk 2 above)
Comment (Hazel-Datastax) [reply src/main/java/io/stargate/sgv2/jsonapi/config/constants/VectorConstant.java]: move `SUPPORTED_SOURCE_MODELS` map from `VectorConstants` to `SourceModel`

File: src/main/java/io/stargate/sgv2/jsonapi/config/constants/DocumentConstants.java [hunk 12]
```diff
@@ -27,9 +27,6 @@ interface Fields {
     /** Document field name that will have text value for which vectorize method in called */
     String VECTOR_EMBEDDING_TEXT_FIELD = "$vectorize";
 
-    /** Key for vector function name definition in cql index. */
-    String VECTOR_INDEX_FUNCTION_NAME = "similarity_function";
-
```
Comment (Hazel-Datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/config/constants/DocumentConstants.java:32]: Duplicated with `VectorConstants`. I think it's better in there, so I remove the one in here

File: src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/VectorConfig.java [hunk 13]
```diff
@@ -67,22 +69,32 @@ public static VectorConfig from(TableMetadata tableMetadata, ObjectMapper object
...
                 }
-                return null;
+                return new AbstractMap.SimpleEntry<>(similarityFunction, sourceModel);
               });
 
       // if now index, or we could not work out the function, default
-      var similarityFunction = indexFunction.orElse(SimilarityFunction.COSINE);
+      var similarityFunction =
+          indexFunction.map(Map.Entry::getKey).orElse(SimilarityFunction.COSINE);
+      var sourModel = indexFunction.map(Map.Entry::getValue).orElse(SourceModel.OTHER);
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/cqldriver/executor/VectorConfig.java:97]: typo: sourModel -> sourceModel ?

File: src/main/java/io/stargate/sgv2/jsonapi/service/schema/SourceModel.java [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+  COHERE_V3("cohere-v3"),
+  GECKO("gecko"),
+  NV_QA_4("nv-qa-4"),
+  OPENAI_V3_LARGE("openai-v3-large"),
+  OPENAI_V3_SMALL("openai-v3-small"),
+  OTHER("other"),
+  UNDEFINED("undefined");
+
+  private final String name;
+  public static final Map<String, SourceModel> SOURCE_MODEL_NAME_MAP =
```
Comment (tatu-at-datastax) [inline src/main/java/io/stargate/sgv2/jsonapi/service/schema/SourceModel.java:25]: Would be better to be `private` and only access via "get" or "find" method for lookups.
Unless the whole `Map` is needed for some reason?

**Review by tatu-at-datastax — APPROVED** [review]
Looks good: two smaller suggestions, approving, you can consider whether to follow suggestions or not.
